Oprava zprovoznění aplikace + refaktorizace

castamir opened the pull request:
- opraveny chybějící složky `www/webtemp/` a `temp/` a patřičným způsobem i `.gitignore`
- opraven konfigurační soubor - opravdu není potřeba mít přihlašovací údaje k produkčnímu prostředí na lokále
- oprava překlepu ve webloaderu
- oprava .htaccess umožňující mj. deploy do podsložky
- přepsána celá aplikační struktura na balíčkovací (třídy jsou organizovány tématicky, nikoliv typově)
- změna jmenného prostoru - nyní vše v namespace `Supervizor`
- konfigurace rozšíření (extensions) vyčleněna do samostatných souborů (snadné "zapnutí" nebo "vypnutí" extension)
- webloader přepsán na extension
- přidán decorator extension (zatím jen na automatické tagování konzolových příkazů)
- konfigurace služeb přesunuta do `config.neon` + zjednodušení syntaxe zápisu
- úklid v presenterech - veškerá logika mimo presentery (ve fasádních třídách)
- v readme jsou informace o právech zápisu
- bower instaluje automaticky bootstrap fonty do `www/fonts`
- rozšířena konfigurace rounteru - nyní je možné explicitně zapnout nebo vypnout ssl
- rozšířena konfigurace databázového připojení o driver

---

**připraveno k review**

Commit:
.gitignore + missing webtemp/ and  temp/ fixed

Commit:
.htaccess removed RewriteBase \ rule

Commit:
config.local.neon - only 1 configuration settings is required at a time

Commit:
webloader: typo in included js

Commit:
namespace + file structure refactoring

Commit:
presenter namespace and mapping

Commit:
config: components.neon merged into services.neon

Salamek commented:
Noo ta struktura... se mi ted moc nelibi... nemuzu v tom najit prakticky vubec nic, proc je to takovy "mišmaš" ? hlavne to co je v app/Supervizor/ Entity smichane se zbytkem ... a tak podobne, pred tim to bylo ve slozce model a clovek vedel kde to hledat, ted je to vsude mozne... a prejmenovani Fixtures tak ze se taky ztratily kamsi kdesi...

Salamek commented on app/bootstrap.php:
```diff
@@ -21,6 +21,6 @@
 // Create Dependency Injection container from config.neon file
 $configurator->addConfig(__DIR__ . '/config/config.neon');
-$configurator->addConfig(__DIR__ . '/config/config.local.neon', \Nette\Configurator::AUTO);
+$configurator->addConfig(__DIR__ . '/config/config.local.neon', Nette\Configurator::NONE);
```
tady neni treba uvadet druhy parametr vubec v tom pripade...

castamir commented:
Protože bylo všechno na jedné hromadě, i to, co tam vůbec nepatří (viz `_fixtures`, které patří spíš do `Console\Commands`).

A není to ještě dodělané. Včera odpoledne jsem opravil chyby pro zprovoznění a večer jsem se vrhnul na refaktorizaci. A jelikož jsem vycházel z těch oprav, tak jsem to přidal i sem.

Commit:
config refactored,  refactored

Commit:
bootstrap: addConfig() removed redundant argument

Salamek commented:
@castamir no jeste jsem se na to dival... a snazil se v tom najit nejakou vysi logiku... a nejak jsem ji nenasel :-D 

projdu to:

- app/* : ok proc ne, i kdyz zbytecna slozka protoze tohle asi nikdy vic jako Supervizor nebude
- app/Supervizor/ : tak jo, uplne zmizelo rozdeleni podle typu, takze kdyz klikam na slozku nebo na soubor tak nemam vlastne paru co tam najdu...

- app/Supervizor/Auth/ : tak ted vidim nejake soubory, Controls - jasne to je pochopitelne, ted me napada desuplna otazka, to bude ve vsech app/Supervizor/*/ (krome Application, Console, Routing, UI, Utils) dalsi slozka Controls ? jak poznam ze [Application, Console, Routing, UI, Utils] jsou neco jineho nez [Auth, Budget, Import, Invoice, Settings, Suppliers] ? Krome toho ze to budu muset proklikat a zapamatovat si to... protoze nazvy ani struktura nerikaji o co goo... no ale zpatky k Auth, tam mame implementaci Nette\Security\IAuthenticator v souboru Authentificator.php ok, to jde... PasswordManager.php, to jeste tak nejak mluvi samo za sebe ale potom User.php ? Entity! Wut ? vsechno v jednom... 

- fixtures maji mnohem dal ke commands nez k models,  commands proste jen spousti prikazy, jejichz kod muze byt prakticky kdekoliv v aplikaci a tento kod nemusi byt spousten jen commandem, ale fixtures plni databazi defaultnimi daty, a fixtures se vzdy drzely blizko modelu... a jmeno Defaults... jsou to Fixtures... cela planeta tomu rika fixtures...

app/Supervizor/Application a tady zacina uplna sranda, mame tu smichane presentery s latte sablonami a to tak ze @layout.latte je ve slozce spolecne s presentery! a hned vedle jsou slozky s jednotlivymi sablonami pro presentery...  a hlavne jsi musel overridnout formatLayoutTemplateFiles, formatTemplateFiles, ja myslim ze kdyz je to projekt postaven na nette tak by se mel chovat jako normalni nette project at to pripadne prispevatele neplete a ne tam vymyslet nejaka vlastni komba...

No nebudu se tady nejak dal extra vykecavat... jen jsem chtel rict ze mi rozdeleni podle typu prijide logictejsi protoze hned vim co v slozce models, presenters, templates, etc. asi najdu... mozna je v tvem navrhnu nejaka skryta krasa... ale ja ji fakt neobevil a to jsem se snazil :-) No stejne zalezi co se bude libit vic cloveku co bude mergovat tohle PR :-) Jinak ostatni upravy jsou ok, jen tak struktura me teda ze zidle nevytrhla :)

castamir commented:
- `app/` dokážu si dost dobře představit, že vedle toho bude nějaká další služba nebo API (třeba react)
- rozdělení je velice podobné, jako jsi měl v modelu - vlastně to podstatné má úplně stejnou adresářovou strukturu (ve složce Budget máš entity a repozitář Budgetu), takže tohle moc nechápu. Jestli se ti stávající struktura nelíbila, proč jsi ji tak udělal? Ve tvých modelových třídách jsem udělal změnu v podstatě jen v namespace, které již není jednotné podle tepu třídy, ale podle tématické oblasti. Výhodou tohoto přístupu je, že nebudu mít jednu velkou složku plnou úplně všech entit a repozitářů a obdobně nebudu mít jednu velkou složku pro všechny komponenty apod. Adresářová struktura bude mít na druhou stranu větší hloubku, což by nemělo vadit, protože mi to vynahradí modularita. Prostě za ty roky, co dělám v Nette, se mi tato organizace projektu osvědčila ze všech nejvíce. A entita `User`? Implementuje `Nette\Security\IIdentity`, takže podle mě není, co řešit. Patří tam, kde je. Můžeme se bavit o tom, zda se její jmenný prostor nemá jmenovat třeba `Supervizor\Security` (a klidně může). Ostatně přejmenování se nebráním.
- Commands by měly volat nějakou logiku a ošetřit její výstupy. Situace je bohužel taková, že tyto třídy sami obsahují spoustu logiky (stejně jako presentery) a jsou tedy vhodným kandidátem pro další refaktorizaci. K pojmenování... celá planeta tomu tak neříká, ale opět se nebráním přejmenování (ale prvně je to potřeba oddělit na logiku a konzolové příkazy)
- tady budu asi nejvíc nesouhlasit. Nette Framework (ani jiný framework) by ti neměl diktovat, jak máš psát. Je to jen pomocník, nástroj, který můžeš použít, ale můžeš ho použít tak, jak vyhovuje tobě. Kdyby se musely presentery používat jen tak, jak je v sandboxu, proč by pak existoval tak obecný router, proč by pak existovalo mapování? Dyť ani nemusíš používat presenter dědící od `Nette\Application\UI\Presenter`. Sám třeba využíváš vlastní mapování Entity na IIdentity. Je to úplně stejný případ úpravy jako jsem použil já. Nette ti nikdy nevnucovalo jediné možné řešení pro daný problém a tohle je naopak využití jedné z možností Nette. K tomuto tématu se ještě jednou vrátím a zkusím vysvětlit, proč jsem zvolil tuhle adresářovou strukturu, ale bude to delší článek...

Každopádně ti tvůj názor neberu. Respektuju ho, ikdyž nemusím souhlasit.

castamir commented:
K adresářové struktuře:

Mějme aplikaci, která má několik modulů (Admin, Front, Intranet, ...). V presenterům a šablonám pak mám přistupovat jak?
1) každý modul bude mít složky `templates` a `presenters`, což povede mj. k následujícím:
   - dědění layoutu bude stylem `{layout '../../templates/@layout'}` 
   - adresářová struktura pro přístup k presenteru uvnitř modulu bude mít hloubku 4 `app/modules/Front/presenters/HomepagePresenter` v případě 1 modulu. V případě zanoření modulů `app/modules/Front/modules/Next/presenters/HomepagePresenter` nám to po 2 roste dál a dál, přitom obsah těchto složek je pomíchaný úplně stejně (vedle sebe jsou složky pro šablony, presentery i moduly, ale je to stejně zbytečná úroveň)

2) moduly budou zcela nezávislé - bude hierarchie složek šablon a hierarchie složek s presentery (nutné přepsat mapování šablon)
  - výhodou je, že 1:1 může celou strukturu šablon nahradit jinou
  - nevýhodou je, že to mapování bude opravdu hodně šílené a hlavně práce s takovými presentery a šablonami bude hodně matoucí a chybová (šablona vytvořená do špatné úrovně zanoření)

3) zrušit dělení na složky templates a presenters a mít presenter a šablony na stejné úrovni a ve stejné složce s tím, že název šablony je prefixován názvem presenteru.
  - vede k přehledné adresářové struktuře, ale je to prakticky nepoužitelné, neboť daná složka bude obsahovat řádově desítky souborů (stejný problém jako s univerzální složkou modelu - příliš velká šířka)
  - výhodou je třeba dědičnost layoutů - vždy o jednu úroveň výš

4) zrušit dělení na složky templates a presenters a umístit složky se šablonami jednotlivých presenterů vedle těchto presenterů (mnou zvolené řešení)
  - šablony jsou dostatečně blízko presenterům, ale nezavazí (krásně se dají generovat šablony společně s presentery z jednoho místa)
  - dědění layoutů je vždy stejné (opět o jednu úroveň výš)
  - submoduly řešeny stejně jako složky se šablonami - tohle řešení snadno odhalí problém stejně pojmenovaných submodulů a presenterů (nicméně jejich společné existenci nijak nebrání)

-------------------

Tady je to hodně otázka vlastních preferencí a priorit. Já mám velmi často aplikace plné různých mdulů a submodulů, takže ocením jednoduchou a snadno generovatelnou adresářovou strukturu (já vlastně většinu věcí generuji, ať už jsou to presentery, šablony, celé kostry modulů, komponenty, entity, atd...), takže poslední varianta mi přijde nejvhodnější.

castamir commented:
Aby to nebylo jem slohové cvičení, tak mám pár návrhů:

- `app/Supervizor/Auth` přejmenovat na `app/Supervizor/Security`
- zcela odstranit `app/Supervizor/Settings` - obsahuje prázkou třídu, která se nikde nevolá ani nepoužívá (nebo jsem to aspoň nikde nenašel)
- přejmenovat stávající `app/Supervizor/Utils` na `app/Supervizor/DI`, neboť obsahuje právě a jen jedno DI rozšíření

Commit:
router: secured ssl flag refactored - mved to config.local.neon
Supervizor\Utils => Supervizor\DI
Supervizor\Auth => Supervizor\Security
removed unused SupervizorSettings

Commit:
presenter: extracted filters into separate class + registration via config

Commit:
cleanup

Commit:
layout: default css/js block content

Commit:
renamed DefaultData to Fixtures

Commit:
bower: post install script to copy bootstrap fonts

Commit:
AjaxPreseter: refactored

Commit:
SupplierFacade: added missing Cache rules

castamir commented:
Vyčistil jsem `AjaxPresenter`. Logika získávání dat je nyní mimo presenter - ve fasádách.
Pár poznámek k původnímu kódu (škoda, že jsem ty poznámky nedělal průběžně a ještě větší škoda, že jsem neudělal review #11 ještě před jeho přijetím):

- Výjimka typu `\Nette\Application\BadRequestException` se nemá vyhazovat přímo, ale místo toho se má volat `$this->error()` podobně, jak je tomu v jiné části tohoto presenteru.
- Logika a výpočty do presenterů nepatří, vzniká pak totiž nepřehledná špageta
  - cache přesunut do služby
  - kontrola modifikace (+ volání `terminate()`) nesouvisí se získáváním dat z cache - odděleno
  - [DRY](https://cs.wikipedia.org/wiki/Don't_repeat_yourself) viz odesílání payload
  - konstantám v presenteru se pokud možno vyhýbat - jejich použitím ti vzniká závislost na presenteru, což je nežádoucí (neříkám tím, že aktuální umístění v `SupplierFacade` je ideální, ale je lepší, než v presenteru)
- Dobrý nápad s `NotModified` response :+1:

Commit:
ErrorPresenter: replaced deprecated Debugger

Commit:
InvoicePresenter: refactored

castamir commented:
Zodpovědnost za výpočet celkové částky v rámci rozpočtové skupiny je vhodnější dát samotné třídě rozpočtové skupiny (`BudgeGroup`), neboť má k dispozici všechny potřebné informace (všechny její položky a tím pádem všechny dílčí částky). Jinak opět výpočet do presenteru nepatří.

Salamek commented:
@castamir 
No strukturu nebu ted resit, ne ze bych nechtel ale bohuzel na to nemam prostor (moc RL prace - hodne casu jsem ztratil jen timhle prepisem)

s pripominkami souhlasim, neslo mi o vytvoreni kompletne vysperkovane aplikace, ale o co nejrychlejsi sprovozneni (neprepisoval jsem to z nudy, ale z nutnosti ze to bylo treba co nejrychleji prepsat, drive nez to naklade vajicka :-D )... a predani komunite at si to muze doladit (coz prave ted delas :-) ). Puvodne jsem si s tim chtel vyhrat vice (chtel jsem zrefactorovat JS a prepsat nazvy promenych do Anglictiny a tuny dalsich veci, pak se vratit k PHP a projit co by slo vycistit), ale jak jsem videl ze lide zasilaji PR / Issues ke staremu kodu tak jsem to dal do funkcniho stavu a poslal PR.
Ano `app/Supervizor/Settings` se muze odstranit, puvodne jsem si myslel ze se bude hodit, nakonec byla k nicemu, takovych veci se tam mozna jeste par najde, stejne jako vycisteni presenteru a jine veci...

castamir commented:
@Salamek v pořádku. Bez tvého přepisu bych do toho nešel :D To nejpracnější jsi udělal, za což klobouk dolů.

Do logiky se snažím nezasahovat (na to by opravdu bylo nutné prvně napsat testy...). Projdu zbývající presentery a asi to už to nechám být.

Commit:
SearchPresenter: refactored, fixed pagination

castamir commented:
Opravil jsem stránkování ve výsledcích vyhledávání. Nejsem zrovna kovaný v Doctrine, tak prosím o [kontrolu](https://github.com/castamir/Supervizor/blob/patch-1/app/Supervizor/Invoice/InvoiceRepository.php#L121-L174), zda dělám stránkování správně. Získávání všech výsledků a ruční výběr stránky totiž asi nebyl úplně efektivní.

Commit:
SignPresenter: failed login message

Commit:
StatisticsPresenter: refactored

Commit:
readme updated

castamir commented:
Poslední presenter vyčištěn. Readme aktualizováno.

Prosím o review, zda mi něco podstatného neuniklo.

Salamek commented on app/Supervizor/Budget/BudgetFacade.php:
```diff
@@ -0,0 +1,117 @@
+                $group['x'] = $u->getX();
+                $group['y'] = $u->getY();
+                $group['barva'] = $u->getColor();
+                $group['max_uhrazeno_udt'] = (new \DateTime)->modify('+2 months')->getTimestamp(); //!FIXME
```
OOPS, ted se divam ze jsem zde zapomel !FIXME, neni to spravne protoze tam je fixni rozsah, spravne by v:
- $group['max_uhrazeno_udt'] mel byt timestamp posledniho uhrazeni faktury v budgetGroup
- $group['min_uhrazeno_udt'] mel byt timestamp prvniho uhrazeni faktury v budgetGroup

tyhle rozsahy se pouzivaji v detailu budgetGroup jako min/max pro filtrovani dle datumu

Salamek commented:
@castamir 
* strankovani (kod) vypada ok
* jen jeden koment na zapomenute //!FIXME kde jsem zapomel pouzit realna data
* SecuredPresenter/SignPresenter je vlastne ted take nepouzivany, puvodne jsem chtel pouzit prihlasovani pro import dat, ale nakonec jsem nechal import jen pres command, ale predpokladam ze stejne nekdo udela nejakou administraci na konfiguraci zdroju importu + zobrazeni statistik a administraci budgetGroup, budgetItem etc, takze asi nicemu nevadi kdyz se tam tyto soubory nechaji

Takze za me :+1:

castamir commented:
@Salamek díky

Mohl bys prosím ty 2 řádky dopsat? Nejsem si totiž úplně jistý, co jsi tím prvním a posledním záznamem myslel.

Jinak těch FIXME jsem tam viděl povíc...

Salamek commented:
@castamir no bohuzel to neotestuji tak nebudu delat PR protoze mi tvoje verze z nejakeho duvodu zhazuje PHP 7 (takze je to na tobe):
```bash
[702157.353197] php-fpm[8027]: segfault at 7fffdb2e7ff0 ip 00007f4af45dc672 sp 00007fffdb2e7ff0 error 6 in libpcre.so.1.2.6[7f4af45b5000+6f000]
[702171.216040] php-fpm[7847]: segfault at 7fffdb2e7ff0 ip 00007f4af45dc672 sp 00007fffdb2e7ff0 error 6 in libpcre.so.1.2.6[7f4af45b5000+6f000]
[702354.296846] php-fpm[8013]: segfault at 7fffdb2e7ff0 ip 00007f4af45dc672 sp 00007fffdb2e7ff0 error 6 in libpcre.so.1.2.6[7f4af45b5000+6f000
```

jinak tady je snippet kodu kde jsou opravene ty !FIXME:
```php
public function getBudgetGroups($budgetGroupId = null)
    {
        $key = 'BudgetGroup' . $budgetGroupId;
        $fallback = function (& $dependencies) use ($budgetGroupId) {
            $dependencies[Cache::EXPIRE] = CacheStorage::EXPIRATION;

            $finalArray = [];
            $filter = [];
            $min = 0;
            $max = 0;
            $total = 0;
            if (!is_null($budgetGroupId)) {
                $filter['id'] = $budgetGroupId;
            }

            /** @var ArrayCollection $s */
            $s = $this->budgetGroupRepository->getBudgetGroupRepository()->findBy($filter);

            /** @var BudgetGroup $u */
            foreach ($s AS $u) {
                $group = [];
                $group['id'] = $u->getSlug();
                $group['nazev'] = $u->getName();
                $group['popis'] = $u->getDescription();
                $group['x'] = $u->getX();
                $group['y'] = $u->getY();
                $group['barva'] = $u->getColor();
                $group['max_uhrazeno_udt'] = 0;
                $group['min_uhrazeno_udt'] = 0;
                $group['objem'] = 0;
                $group['pocet'] = 0;
                $group['polozky'] = [];
                foreach ($u->getBudgetItems() AS $i) {
                    $item = [];
                    $item['id'] = $i->getIdentifier();
                    $item['nazev'] = $i->getName();
                    $amount = 0;
                    /** @var InvoiceItem $invoiceItem */
                    foreach ($i->getinvoiceItems() AS $invoiceItem) {
                        $amount += $invoiceItem->getAmount();
                        $paidTimestamp = $invoice->getInvoice()->getPaid()->getTimestamp();

                        $group['max_uhrazeno_udt'] = max($group['max_uhrazeno_udt'], $paidTimestamp);
                        if (!$group['min_uhrazeno_udt'])
                        {
                            $group['min_uhrazeno_udt'] = $paidTimestamp;
                        }
                        else
                        {
                            $group['min_uhrazeno_udt'] = min($group['min_uhrazeno_udt'], $paidTimestamp);
                        }
                    }
                    $group['objem'] += $item['objem'] = $amount;
                    $group['pocet'] += $item['pocet'] = $i->getinvoiceItems()->count();
                    $item['skupina_id'] = $u->getId();
                    $group['polozky'][] = $item;

                    $objem = [];
                    foreach ($group['polozky'] as $key => $row) {
                        $objem[$key] = $row['objem'];
                    }
                    array_multisort($objem, SORT_DESC, $group['polozky']);
                }

                $max = max($group['objem'], $max);

                if (!$min) {
                    $min = $group['objem'];
                } else {
                    $min = min($group['objem'], $min);
                }

                $total += $group['objem'];
                $finalArray[$u->getSlug()] = $group;

                $objem = [];
                foreach ($finalArray as $key => $row) {
                    $objem[$key] = $row['objem'];
                }
                array_multisort($objem, SORT_DESC, $finalArray);
            }

            return [
                'skupiny' => $finalArray,
                'stats' => [
                    'min' => $min,
                    'max' => $max,
                    'total' => $total
                ]
            ];
        };
        
        return $this->cacheStorage->load($key, $fallback);
    }
```

Vsechny ostatni !FIXME se netykaji funkcnosti ale vylepseni, napr typy faktur do constant a pouzivat int v DB, a jine veci...

Commit:
BudgetFacade::getBudgetGroups() fixed range

castamir commented:
opravil jsem jeden překlep:

`$paidTimestamp = $invoice->getInvoice()` (nedefinovaná proměnná `$invoice`)
na
`$paidTimestamp = $invoiceItem->getInvoice()`

na PHP7 se mrknu až se vyspím ;)

Salamek commented:
@castamir
Jasny :)

 jeste jsou !FIXME v https://github.com/castamir/Supervizor/blob/patch-1/app/Supervizor/Application/Invoice/default.latte#L55

tam patri (asi, z puvodniho kodu jsem moc nepochopil co ze se to ma vlastne do tech konstant rvat z puvodni konfigurace importu v jsonu... nebo to tam uplne chybelo)

```html
<td id="odberatel">
    <h3>Odběratel</h3>
    <h4>{$invoice->getImport()->getImportGroup()->getName()}</h4>
    <p>{* !FIXME add description to ImportGroup ? $invoice->getImport()->getImportGroup()->getDescription()*}</p>
</td>
```
(Nektere gettery budou mozna chybet v entitach)

Commit:
webloader: fixed base path

config: driver config

Commit:
importer:import:all fixed json loading

Commit:
fixed bug in entity relations and mapping

castamir commented:
Rozjel jsem to na PHP 7.0.4, ale pouze na Windows (linux jsem nezkoušel - momentálně ho nepoužívám). Nějaké drobné chybky tam byly, ale nemyslím si, že nějak zásadní. Každopádně doporučuji smazat cache (upravoval jsem mapování entit).

Salamek commented on app/Supervizor/Budget/BudgetGroup.php:
```diff
@@ -63,7 +62,7 @@ class BudgetGroup extends Nette\Object
     /**
      * @var ArrayCollection|BudgetItem[]
-     * @ORM\OneToMany(targetEntity="BudgetItem", mappedBy="budgetGroup",cascade={"persist"})
+     * @ORM\OneToMany(targetEntity="\Supervizor\Budget\BudgetItem", mappedBy="budgetGroup",cascade={"persist"})
```
Tady by melo stacit jen `BudgetItem` a dat `Supervizor\Budget\BudgetItem` do `use` stejne jako v ostatnich entitach...

Salamek commented:
@castamir ok, tak jsem to tak zbezne prosel a zda se ze vse jede (PHP 7.0.5)

SmallhillCZ commented:
Díky všem, kouknu na to zas o víkendu...
